File: AmigurumiMaker/app/src/main/java/com/example/amigurumimaker/domain/GeometryComputer.kt
package com.example.amigurumimaker.domain

import com.example.amigurumimaker.domain.model.*
import kotlin.math.floor
import kotlin.math.sqrt

/**
 * Computes [BlockGeometry] (vertices, edges, face indices) for a list of [RowDef]s.
 *
 * - Splits RAMP_LEFT tokens into RAMP_LEFT vs RAMP_RIGHT based on block position.
 * - Centers each row at x=0.
 * - Y = rowNumber - 1 (0-indexed).
 * - Z = 0 for all blocks.
 */
object GeometryComputer {

    fun compute(rowDefs: List<RowDef>): List<BlockGeometry> {
        val geometries = mutableListOf<BlockGeometry>()

        for (rowDef in rowDefs) {
            val expanded = RowExpander.expand(rowDef)
            for (singleRow in expanded) {
                val rowNumber = singleRow.startRow
                val totalBlocks = singleRow.tokens.sumOf { it.count }
                val startX = -floor(totalBlocks / 2f)

                var blockIndex = 0
                for (token in singleRow.tokens) {
                    for (i in 0 until token.count) {
                        val cx = startX + blockIndex
                        val cy = (rowNumber - 1).toFloat()
                        val cz = 0f

                        val actualType = if (token.type == TokenType.RAMP_LEFT) {
                            if (blockIndex < totalBlocks / 2f) TokenType.RAMP_LEFT else TokenType.RAMP_RIGHT
                        } else {
                            token.type
                        }

                        val geometry = when (actualType) {
                            TokenType.CUBE -> cubeGeometry(cx, cy, cz)
                            TokenType.RAMP_LEFT -> rampLeftGeometry(cx, cy, cz)
                            TokenType.RAMP_RIGHT -> rampRightGeometry(cx, cy, cz)
                            TokenType.PLACEHOLDER -> BlockGeometry(emptyList(), emptyList(), emptyList())
                        }

                        geometries.add(geometry)
                        blockIndex++
                    }
                }
            }
        }

        return geometries
    }

    private fun cubeGeometry(cx: Float, cy: Float, cz: Float): BlockGeometry {
        val vertices = listOf(
            Float3(-0.5f + cx, -0.5f + cy, -0.5f + cz),  // 0: left-bottom-front
            Float3(0.5f + cx, -0.5f + cy, -0.5f + cz),   // 1: right-bottom-front
            Float3(0.5f + cx, 0.5f + cy, -0.5f + cz),    // 2: right-top-front
            Float3(-0.5f + cx, 0.5f + cy, -0.5f + cz),   // 3: left-top-front
            Float3(-0.5f + cx, -0.5f + cy, 0.5f + cz),   // 4: left-bottom-back
            Float3(0.5f + cx, -0.5f + cy, 0.5f + cz),    // 5: right-bottom-back
            Float3(0.5f + cx, 0.5f + cy, 0.5f + cz),     // 6: right-top-back
            Float3(-0.5f + cx, 0.5f + cy, 0.5f + cz),    // 7: left-top-back
        )

        val edges = listOf(
            IntPair(0, 1), IntPair(1, 2), IntPair(2, 3), IntPair(3, 0),
            IntPair(4, 5), IntPair(5, 6), IntPair(6, 7), IntPair(7, 4),
            IntPair(0, 4), IntPair(1, 5), IntPair(2, 6), IntPair(3, 7),
        )

        val faceIndices = listOf(
            // Front: (0,2,1),(0,3,2)
            IntTriple(0, 2, 1), IntTriple(0, 3, 2),
            // Back: (4,5,6),(4,6,7)
            IntTriple(4, 5, 6), IntTriple(4, 6, 7),
            // Top: (3,2,6),(3,6,7)
            IntTriple(3, 2, 6), IntTriple(3, 6, 7),
            // Bottom: (0,1,5),(0,5,4)
            IntTriple(0, 1, 5), IntTriple(0, 5, 4),
            // Right: (1,2,6),(1,6,5)
            IntTriple(1, 2, 6), IntTriple(1, 6, 5),
            // Left: (0,4,7),(0,7,3)
            IntTriple(0, 4, 7), IntTriple(0, 7, 3),
        )

        return BlockGeometry(vertices, edges, faceIndices)
    }

    /**
     * RAMP_LEFT: triangular prism, diagonal from lower-left to upper-right.
     *
     * Vertices:
     *   0: left-bottom-front  (-0.5, -0.5, -0.5)
     *   1: right-bottom-front (0.5, -0.5, -0.5)
     *   2: right-top-front    (0.5, 0.5, -0.5)
     *   3: left-bottom-back   (-0.5, -0.5, 0.5)
     *   4: right-bottom-back  (0.5, -0.5, 0.5)
     *   5: right-top-back     (0.5, 0.5, 0.5)
     */
    private fun rampLeftGeometry(cx: Float, cy: Float, cz: Float): BlockGeometry {
        val vertices = listOf(
            Float3(-0.5f + cx, -0.5f + cy, -0.5f + cz),  // 0
            Float3(0.5f + cx, -0.5f + cy, -0.5f + cz),   // 1
            Float3(0.5f + cx, 0.5f + cy, -0.5f + cz),    // 2
            Float3(-0.5f + cx, -0.5f + cy, 0.5f + cz),   // 3
            Float3(0.5f + cx, -0.5f + cy, 0.5f + cz),    // 4
            Float3(0.5f + cx, 0.5f + cy, 0.5f + cz),     // 5
        )

        val edges = listOf(
            IntPair(0, 1), IntPair(1, 2), IntPair(2, 0),
            IntPair(3, 4), IntPair(4, 5), IntPair(5, 3),
            IntPair(0, 3), IntPair(1, 4), IntPair(2, 5),
        )

        val faceIndices = listOf(
            IntTriple(0, 1, 2),  // Front
            IntTriple(3, 5, 4),  // Back
            IntTriple(0, 1, 4), IntTriple(0, 4, 3),  // Bottom
            IntTriple(1, 2, 5), IntTriple(1, 5, 4),  // Right
            IntTriple(2, 0, 3), IntTriple(2, 3, 5),  // Diagonal
        )

        return BlockGeometry(vertices, edges, faceIndices)
    }

    /**
     * RAMP_RIGHT: triangular prism, diagonal from lower-right to upper-left.
     *
     * Vertices:
     *   0: left-bottom-front  (-0.5, -0.5, -0.5)
     *   1: right-bottom-front (0.5, -0.5, -0.5)
     *   2: left-top-front     (-0.5, 0.5, -0.5)
     *   3: left-bottom-back   (-0.5, -0.5, 0.5)
     *   4: right-bottom-back  (0.5, -0.5, 0.5)
     *   5: left-top-back      (-0.5, 0.5, 0.5)
     */
    private fun rampRightGeometry(cx: Float, cy: Float, cz: Float): BlockGeometry {
        val vertices = listOf(
            Float3(-0.5f + cx, -0.5f + cy, -0.5f + cz),  // 0
            Float3(0.5f + cx, -0.5f + cy, -0.5f + cz),   // 1
            Float3(-0.5f + cx, 0.5f + cy, -0.5f + cz),   // 2
            Float3(-0.5f + cx, -0.5f + cy, 0.5f + cz),   // 3
            Float3(0.5f + cx, -0.5f + cy, 0.5f + cz),    // 4
            Float3(-0.5f + cx, 0.5f + cy, 0.5f + cz),    // 5
        )

        val edges = listOf(
            IntPair(0, 1), IntPair(1, 2), IntPair(2, 0),
            IntPair(3, 4), IntPair(4, 5), IntPair(5, 3),
            IntPair(0, 3), IntPair(1, 4), IntPair(2, 5),
        )

        val faceIndices = listOf(
            IntTriple(0, 1, 2),  // Front
            IntTriple(3, 5, 4),  // Back
            IntTriple(0, 1, 4), IntTriple(0, 4, 3),  // Bottom
            IntTriple(0, 2, 5), IntTriple(0, 5, 3),  // Left
            IntTriple(1, 2, 5), IntTriple(1, 5, 4),  // Diagonal
        )

        return BlockGeometry(vertices, edges, faceIndices)
    }
}

File: amigurumiMakerType/src/domain/GeometryComputer.ts
// src/domain/GeometryComputer.ts
// Port of Kotlin GeometryComputer — 2D block geometry (cubes, ramps)

import { BlockGeometry, Float3 } from './model/models';

type TokenType = 'CUBE' | 'RAMP_LEFT' | 'RAMP_RIGHT' | 'PLACEHOLDER';

interface RowToken {
  type: TokenType;
  count: number;
}

/**
 * Compute block geometries for a list of token rows.
 * Each token produces cube, ramp_left, or ramp_right geometry.
 */
export function compute(tokens: RowToken[]): BlockGeometry[] {
  const geometries: BlockGeometry[] = [];

  for (const row of tokens) {
    let blockIndex = 0;
    const totalBlocks = row.count;

    for (let i = 0; i < row.count; i++) {
      const cx = -Math.floor(totalBlocks / 2) + blockIndex;
      const cy = 0; // Single row, y=0
      const cz = 0;

      let actualType = row.type;
      // For ramp types, split by position relative to center
      if (row.type === 'RAMP_LEFT' || row.type === 'RAMP_RIGHT') {
        actualType = blockIndex < totalBlocks / 2 ? 'RAMP_LEFT' : 'RAMP_RIGHT';
      }

      let geometry: BlockGeometry;
      switch (actualType) {
        case 'CUBE':
          geometry = cubeGeometry(cx, cy, cz);
          break;
        case 'RAMP_LEFT':
          geometry = rampLeftGeometry(cx, cy, cz);
          break;
        case 'RAMP_RIGHT':
          geometry = rampRightGeometry(cx, cy, cz);
          break;
        default:
          geometry = { vertices: [], edges: [], faceIndices: [] };
          break;
      }

      geometries.push(geometry);
      blockIndex++;
    }
  }

  return geometries;
}

/**
 * Cube: 8 vertices, 12 edges, 12 face indices (6 faces × 2 triangles)
 */
function cubeGeometry(cx: number, cy: number, cz: number): BlockGeometry {
  const vertices: Float3[] = [
    { x: -0.5 + cx, y: -0.5 + cy, z: -0.5 + cz }, // 0: left-bottom-front
    { x: 0.5 + cx, y: -0.5 + cy, z: -0.5 + cz },  // 1: right-bottom-front
    { x: 0.5 + cx, y: 0.5 + cy, z: -0.5 + cz },   // 2: right-top-front
    { x: -0.5 + cx, y: 0.5 + cy, z: -0.5 + cz },  // 3: left-top-front
    { x: -0.5 + cx, y: -0.5 + cy, z: 0.5 + cz },  // 4: left-bottom-back
    { x: 0.5 + cx, y: -0.5 + cy, z: 0.5 + cz },   // 5: right-bottom-back
    { x: 0.5 + cx, y: 0.5 + cy, z: 0.5 + cz },    // 6: right-top-back
    { x: -0.5 + cx, y: 0.5 + cy, z: 0.5 + cz },   // 7: left-top-back
  ];

  const edges: [number, number][] = [
    [0, 1], [1, 2], [2, 3], [3, 0],
    [4, 5], [5, 6], [6, 7], [7, 4],
    [0, 4], [1, 5], [2, 6], [3, 7],
  ];

  const faceIndices: [number, number, number][] = [
    // Front
    [0, 2, 1], [0, 3, 2],
    // Back
    [4, 5, 6], [4, 6, 7],
    // Top
    [3, 2, 6], [3, 6, 7],
    // Bottom
    [0, 1, 5], [0, 5, 4],
    // Right
    [1, 2, 6], [1, 6, 5],
    // Left
    [0, 4, 7], [0, 7, 3],
  ];

  return { vertices, edges, faceIndices };
}

/**
 * RAMP_LEFT: triangular prism, diagonal from lower-left to upper-right.
 * 6 vertices, 9 edges, 8 face indices.
 */
function rampLeftGeometry(cx: number, cy: number, cz: number): BlockGeometry {
  const vertices: Float3[] = [
    { x: -0.5 + cx, y: -0.5 + cy, z: -0.5 + cz }, // 0
    { x: 0.5 + cx, y: -0.5 + cy, z: -0.5 + cz },  // 1
    { x: 0.5 + cx, y: 0.5 + cy, z: -0.5 + cz },   // 2
    { x: -0.5 + cx, y: -0.5 + cy, z: 0.5 + cz },  // 3
    { x: 0.5 + cx, y: -0.5 + cy, z: 0.5 + cz },   // 4
    { x: 0.5 + cx, y: 0.5 + cy, z: 0.5 + cz },    // 5
  ];

  const edges: [number, number][] = [
    [0, 1], [1, 2], [2, 0],
    [3, 4], [4, 5], [5, 3],
    [0, 3], [1, 4], [2, 5],
  ];

  const faceIndices: [number, number, number][] = [
    [0, 1, 2],              // Front
    [3, 5, 4],              // Back
    [0, 1, 4], [0, 4, 3],  // Bottom
    [1, 2, 5], [1, 5, 4],  // Right
    [2, 0, 3], [2, 3, 5],  // Diagonal
  ];

  return { vertices, edges, faceIndices };
}

/**
 * RAMP_RIGHT: triangular prism, diagonal from lower-right to upper-left.
 * 6 vertices, 9 edges, 8 face indices.
 */
function rampRightGeometry(cx: number, cy: number, cz: number): BlockGeometry {
  const vertices: Float3[] = [
    { x: -0.5 + cx, y: -0.5 + cy, z: -0.5 + cz }, // 0
    { x: 0.5 + cx, y: -0.5 + cy, z: -0.5 + cz },  // 1
    { x: -0.5 + cx, y: 0.5 + cy, z: -0.5 + cz },  // 2
    { x: -0.5 + cx, y: -0.5 + cy, z: 0.5 + cz },  // 3
    { x: 0.5 + cx, y: -0.5 + cy, z: 0.5 + cz },   // 4
    { x: -0.5 + cx, y: 0.5 + cy, z: 0.5 + cz },   // 5
  ];

  const edges: [number, number][] = [
    [0, 1], [1, 2], [2, 0],
    [3, 4], [4, 5], [5, 3],
    [0, 3], [1, 4], [2, 5],
  ];

  const faceIndices: [number, number, number][] = [
    [0, 1, 2],              // Front
    [3, 5, 4],              // Back
    [0, 1, 4], [0, 4, 3],  // Bottom
    [0, 2, 5], [0, 5, 3],  // Left
    [1, 2, 5], [1, 5, 4],  // Diagonal
  ];

  return { vertices, edges, faceIndices };
}
